Added code for e2e tests for network policy.

tathagatachowdhury opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
This PR adds e2e test cases for network policies to test the communication between pods based on pod-selectors or namespace-selector or both.
Total of 8 test cases have been added.
While executing the e2e network policy tests, it was observed that if multiple pods, with same label has to be created in an It() block then there is clash of pod name. To resolve this, the pod name, provided at the time pod creation, is appended with a "-" + "<random string of 5 characters>".

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

dcbw commented:
/assign danwinship

dcbw commented:
Flake?

ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/cluster/images/hyperkube/BUILD:4:1: no such package '@debian-hyperkube-base-amd64//image': Pull command failed: Traceback (most recent call last):

dcbw commented:
Flake?

ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/cluster/images/hyperkube/BUILD:4:1: no such package '@debian-hyperkube-base-amd64//image': Pull command failed: Traceback (most recent call last):

tathagatachowdhury commented:
/retest

danwinship commented:
>  While executing the e2e network policy tests, it was observed that if multiple pods, with same label has to be created in an It() block then there is clash of pod name. 

"clash" how? Like, if the tests are being run in parallel? (Do we do that?)

> To resolve this, the pod name, provided at the time pod creation, is appended with a "-" + "<random string of 5 characters>".

It seems like a simpler fix would be to just make sure each test uses a different pod name...

danwinship commented:
if the pod-base-name thing *is* necessary, it should be done as a separate commit from adding the new tests

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -122,7 +128,56 @@ var _ = SIGDescribe("NetworkPolicy", func() {
 		})
 
-		It("should enforce policy based on NamespaceSelector [Feature:NetworkPolicy]", func() {
+		It("should enforce policy, to allow traffic from pods within a namespace and deny ingress traffic for different namespace, based on PodSelector [Feature:NetworkPolicy]", func() {
```
This could just be added as an additional `testCannotConnect` inside the existing `It("should enforce policy based on PodSelector [Feature:NetworkPolicy]"`

tathagatachowdhury replied:
Acknowledged. I have included this test in the existing It().

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -171,6 +226,85 @@ var _ = SIGDescribe("NetworkPolicy", func() {
 			testCanConnect(f, nsB, "client-b", service, 80)
 		})
 
+		It("should enforce policy, to allow traffic only from a pod in a different namespaces than that of server's namespace,  based on PodSelector and NamespaceSelector [Feature:NetworkPolicy]", func() {
```
the descriptions don't need to be so wordy. I'd just say `It("should enforce policy based on combined NamespaceSelector and PodSelector"`. (Likewise with most of the other new tests; follow the example of the existing tests.)

tathagatachowdhury replied:
I have shortened the description and tried to avoid duplicate descriptions.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+									},
+								},
+							},
+							// Need to allow traffic to DNS for name relsolution
```
typo ("relsolution")

tathagatachowdhury replied:
Acknowledged and updated.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+								},
+								PodSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{
+										"k8s-app": "kube-dns",
```
The OpenShift e2e tests pull in a bunch of the upstream kubernetes e2e tests, but are run in a slightly different environment where `podSelector: { matchLabels: { k8s-app: kube-dns } }` would not be correct. I don't know if anyone else does anything similar...

But anyway, there's an easy workaround for that (which is also more future-proof just within kubernetes): just allow egress to all destinations on port 53.

tathagatachowdhury replied:
Acknowledged and done.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			// client-a should be able to contact server as the policy is to allow all
+			By("Creating client-a which should be able to contact the server.", func() {
+				testCanConnect(f, f.Namespace, "client-a", service, 80)
+			})
```
For completeness you might as well test that client-b can still connect as well. (likewise in the corresponding egress test)

tathagatachowdhury replied:
Acknowledged and added the tests in both ingress and egress.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+						},
+					},
+					PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+					// Allow traffic only to "client-b"
```
Don't you need to allow DNS here as well? (Otherwise the test will fail for the wrong reason...)

tathagatachowdhury replied:
DNS is required and I have added it.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			policyDenyClientA := &networkingv1.NetworkPolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: f.Namespace.Name,
+					Name:      "allow-from-client-b-pod-selector",
```
name is misleading; this is an "allow **to** client b" policy

tathagatachowdhury replied:
Acknowledged and updated the name with "allow-to-server-a-pod-selector".

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			Expect(err).NotTo(HaveOccurred())
+			defer cleanupNetworkPolicy(f, policyAllowAll)
+
+			// client-a should be able to contact server as the policy is to allow all
```
This is wrong; ingress and egress are separate things. The "allow-all" ingress policy says that all pods will *accept* traffic from all other pods, but the egress policy above is still in effect and says that "client-a" can only *send* traffic to "client-b". (In other words, at this point, the server would be happy to accept traffic from client-a if client-a could send that traffic, but client-a cannot send that traffic, so the server never gets the opportunity to accept it.)

Did you test these new tests against some network plugin? If so, and that plugin passed this test, then that plugin is broken. (In which case, hooray for having more tests!)

tathagatachowdhury replied:
Acknowledged and changed the test.

caseydavenport replied:
@danwinship has this comment be addressed now? 

I think it has, just looking at the current code (there isn't any mention of ingress in this test any more and it seems to be doing the right thing).

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+						},
+					},
+					PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+					// Allow traffic from all pods
```
"Allow traffic from client-a"

tathagatachowdhury replied:
Acknowledged and updated.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+				testCanConnect(f, f.Namespace, "client-a", service, 80)
+			})
+
+			cleanupNetworkPolicy(f, policyAllowClientA)
```
need a `By("deleting the network policy allowing traffic from client-a")` here

tathagatachowdhury replied:
Acknowledged and added the By().

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			})
+		})
+
+		It("should enforce policies and test impact of deletion of policy on pod traffic based on PodSelector [Feature:NetworkPolicy]", func() {
```
"test impact of deletion" describes what the test itself is doing, not what it is expecting kubernetes to do.
`It("should stop enforcing policies after they are deleted`

tathagatachowdhury replied:
Acknowledged and updated.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+
+			// Before creation of the policy client-a should be able to connect to server-b
+			By("Creating client-a which should not be able to contact the server-b.", func() {
+				testCanConnect(f, f.Namespace, "client-a", serviceB, 80)
```
which *should* be able to contact

tathagatachowdhury replied:
Acknowledged and updated.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+	Context("NetworkPolicy between pods, each containing a server container and client container", func() {
+		BeforeEach(func() {
+			By("Creating a pod with simple server container, that serves on port 80 and 81, and a client container.")
+			podServer, service = createPodWithClientServerAndService(f, f.Namespace, "pod-a", []int{80, 81}, "client-a", "pod-a-server-responded")
```
you don't need to do that; you can just create a client and a server with the same labels, and you'd be testing the same thing

tathagatachowdhury replied:
I have tried the way you suggest in the comment. I have written this test in this way for two reasons:
1. I created a client and a server with different labels for one test, then deleted them and recreate another set of server and client, with labels exchanged. In this case, I found that there are times when the creation of the second set of server and client starts before the first set is deleted completely. This leads to failure of the test as object being created already exists. This behavior is flaky. Prior to my changes, pod name and pod label were same. So I have made the pod name distinct by adding the random string (this is also the answer for the question you asked earlier).

2. For testing of this scenario, it seemed better to test without deleting the pods.

Please let me know your thoughts.

song-jiang replied:
@danwinship  I would suggest to use exact same pod for connectivity check before and after applying network policy. If we recreate another pair of client/server pods with same label, those pods may be rescheduled onto different nodes and the connectivity between them could be affected by other aspect instead of network policy.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			targetA := fmt.Sprintf("%s.%s", service.Name, service.Namespace)
+			cmdB := fmt.Sprintf("wget -T 2 %v -O - | grep pod-a-server-responded && exit 0 || exit 1", targetA)
+
+			stdoutB, err := framework.RunKubectl(
```
There are wrappers for this. Look at other tests. (Except that also, I don't think you need the combined client-and-server pod, so you don't need to do `kubectl exec`.)

danwinship commented:
also, if you're having fun adding NetworkPolicy tests, the only test left from #46625 that still needs to be written is a test that specifying a `networkingv1.NetworkPolicyPort` with a `Protocol` but no `Port` will allow access to all ports of that protocol (but not of other protocols).

tathagatachowdhury commented on test/e2e/network/network_policy.go:
```diff
@@ -122,7 +128,56 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			nsB, err := f.CreateNamespace(nsBName, map[string]string{
+				"ns-name": nsBName,
+			})
+			Expect(err).NotTo(HaveOccurred())
```
Acknowledged and updated.

tathagatachowdhury commented:
> that still needs to be written is a test that specifying a `networkingv1.NetworkPolicyPort` with a `Protocol` but no `Port` will allow access to all ports of that protocol (but not of other protocols).

Can you please elaborate on the test case you are talking about?

danwinship commented:
Per the API docs, if you have a NetworkPolicy like:

    kind: NetworkPolicy
    apiVersion: networking.k8s.io/v1
    metadata:
      name: allow-to-all-tcp
    spec:
      podSelector:
        matchLabels:
          pod-name: blah
      ingress:
      - ports:
        - protocol: TCP

Then the selected pods should accept connections on all TCP ports but not any UDP ports.
(This is a silly feature, but it's in the spec.)

tathagatachowdhury commented:
> 
> 
> Per the API docs, if you have a NetworkPolicy like:
> 
> ```
> kind: NetworkPolicy
> apiVersion: networking.k8s.io/v1
> metadata:
>   name: allow-to-all-tcp
> spec:
>   podSelector:
>     matchLabels:
>       pod-name: blah
>   ingress:
>   - ports:
>     - protocol: TCP
> ```
> 
> Then the selected pods should accept connections on all TCP ports but not any UDP ports.
> (This is a silly feature, but it's in the spec.)

@danwinship 
I am working on this. The UDP server is turning out to be a time consuming to handle as port scan with `nc`  after applying protocol allowing only TCP does not give the correct status for communication. Either I will need to find a way to use `nc` to send back packets to client and then verify communication or I will have to create a new image for UDP server, which will be a another task in itself.
Please let me know if you have any suggestions.

Also as this test case may take sometime, can we please move ahead with the PR and merge the remaining test case in a separate PR? If that is possible then please let me know the issue number against which I can raise a new PR.

danwinship commented:
/remove-area apiserver
/remove-area cloudprovider
/remove-area code-generation
/remove-area conformance
/remove-area dependency
/remove-area e2e-test-framework
/remove-area kubeadm
/remove-area kubectl
/remove-area kubelet
/remove-area release-eng
/remove-kind api-change
/remove-sig api-machinery
/remove-sig apps
/remove-sig architecture
/remove-sig auth
/remove-sig autoscaling
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig contributor-experience
/remove-sig gcp
/remove-sig instrumentation
/remove-sig node
/remove-sig release
/remove-sig scheduling
/remove-sig storage
/remove-sig testing

tathagatachowdhury commented:
> 
> 
> Per the API docs, if you have a NetworkPolicy like:
> 
> ```
> kind: NetworkPolicy
> apiVersion: networking.k8s.io/v1
> metadata:
>   name: allow-to-all-tcp
> spec:
>   podSelector:
>     matchLabels:
>       pod-name: blah
>   ingress:
>   - ports:
>     - protocol: TCP
> ```
> 
> Then the selected pods should accept connections on all TCP ports but not any UDP ports.
> (This is a silly feature, but it's in the spec.)

@danwinship 
I have implemented this test case. Please have a look.

tathagatachowdhury commented:
/retest

danwinship commented:
>     1. I found that there are times when the creation of the second set of server and client starts before the first set is deleted completely. This leads to failure of the test as object being created already exists. This behavior is flaky. Prior to my changes, pod name and pod label were same. So I have made the pod name distinct by adding the random string (this is also the answer for the question you asked earlier).

You don't need to change the label. If you want to guarantee that pods have unique names, you can do

    pod, err := f.ClientSet.CoreV1().Pods(namespace.Name).Create(&v1.Pod{
    	ObjectMeta: metav1.ObjectMeta{
    		GenerateName: podName+"-"
    		Labels: map[string]string{
    			"pod-name": podName,
    		},
    	},
    	...

That will generate a guaranteed-unique `Name`, but still use the provided `podName` in the label. Even if there are old not-fully-deleted pods around and so you have two pods with `labels: { pod-name: server }`, that doesn't matter, because the tests aren't going to do anything with the old pod, so it doesn't matter whether it is selected by the NetworkPolicy or not

danwinship commented:
> > Then the selected pods should accept connections on all TCP ports but not any UDP ports.
> > (This is a silly feature, but it's in the spec.)
> 
> I have implemented this test case. Please have a look.

Hm... this is a huge change... I'd say to revert things back to how it was before, without that test, and file a separate PR afterward for adding that test. (And we probably don't want it to be done so invasively; you should leave the existing methods for creating tcp clients and servers, and have separate methods for UDP servers. And there are already utility methods for doing that elsewhere.)

tathagatachowdhury commented:
> > > Then the selected pods should accept connections on all TCP ports but not any UDP ports.
> > > (This is a silly feature, but it's in the spec.)
> > 
> > 
> > I have implemented this test case. Please have a look.
> 
> Hm... this is a huge change... I'd say to revert things back to how it was before, without that test, and file a separate PR afterward for adding that test. (And we probably don't want it to be done so invasively; you should leave the existing methods for creating tcp clients and servers, and have separate methods for UDP servers. And there are already utility methods for doing that elsewhere.)

@danwinship I have removed the changes.

tathagatachowdhury commented:
/retest

tathagatachowdhury commented:
> > 1. I found that there are times when the creation of the second set of server and client starts before the first set is deleted completely. This leads to failure of the test as object being created already exists. This behavior is flaky. Prior to my changes, pod name and pod label were same. So I have made the pod name distinct by adding the random string (this is also the answer for the question you asked earlier).
> 
> You don't need to change the label. If you want to guarantee that pods have unique names, you can do
> 
> ```
> pod, err := f.ClientSet.CoreV1().Pods(namespace.Name).Create(&v1.Pod{
> 	ObjectMeta: metav1.ObjectMeta{
> 		GenerateName: podName+"-"
> 		Labels: map[string]string{
> 			"pod-name": podName,
> 		},
> 	},
> 	...
> ```
> 
> That will generate a guaranteed-unique `Name`, but still use the provided `podName` in the label. Even if there are old not-fully-deleted pods around and so you have two pods with `labels: { pod-name: server }`, that doesn't matter, because the tests aren't going to do anything with the old pod, so it doesn't matter whether it is selected by the NetworkPolicy or not

@danwinship I have updated the code to use `GenerateName`.

oomichi commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +641,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			framework.ExpectNoError(err, "Error occurred while trying to run \"kubectl exec\" in client-b.")
+			// Trimming leading and trailing spaces and newlines.
+			outputB := strings.TrimSpace(stdoutB)
+			gomega.Expect(outputB).To(gomega.Equal("pod-a-server-responded"))
```
We can use `framework.ExpectEqual(outputB, "pod-a-server-responded")` here instead then we can remove `gomega` import.

oomichi commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +641,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			framework.ExpectNoError(err, "Error occurred while trying to run \"kubectl exec\" in server-a.")
+			// Trimming leading and trailing spaces and newlines.
+			outputA := strings.TrimSpace(stdoutA)
+			gomega.Expect(outputA).To(gomega.Equal("pod-b-server-responded"))
```
ditto

oomichi commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +641,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+					e2elog.Failf("Pod %s should not be able to connect to service %s, but was not able to connect.\n", podB.Name, service.Name)
+				}
+			}
+			gomega.Expect(len(outputB)).To(gomega.Equal(0))
```
ditto: `framework.ExpectEqual(len(outputB), 0)`

oomichi commented:
/cc @oomichi

danwinship commented:
Sorry for not getting back to this sooner. #78177 has been merged now which includes some new tests that may be duplicates with some of these. (The PR needs to be rebased and repushed even if there isn't any duplication.)

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -55,6 +57,9 @@ var _ = SIGDescribe("NetworkPolicy", func() {
 				framework.ExpectNoError(err)
 			})
 
+			// podServerLabelSelector holds the value for the podSever's label "pod-name".
```
(typo "podSever")

tathagatachowdhury replied:
I have corrected typo.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -85,8 +90,29 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			nsBName := f.BaseName + "-b"
+			// The CreateNamespace helper uses the input name as a Name Generator, so the namespace itself
+			// will have a different name than what we are setting as the value of ns-name.
+			// This is fine as long as we don't try to match the label as nsB.Name in our policy.
```
I don't think this comment is necessary (in any of the places where it appears).

tathagatachowdhury replied:
I have removed the comments from all the places in the file.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -351,6 +380,85 @@ var _ = SIGDescribe("NetworkPolicy", func() {
 			testCanConnect(f, nsB, "client-b", service, 80)
 		})
 
+		ginkgo.It("should enforce policy to allow traffic only from a pod in a different namespaces based on PodSelector and NamespaceSelector [Feature:NetworkPolicy]", func() {
```
"in a different namespace" (not "namespaces")

tathagatachowdhury replied:
Updated "namespaces" to "namespace".

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -351,6 +380,85 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			})
+			ginkgo.By("Creating client-a, in server's namespace, which should be able to contact the server.", func() {
+				testCanConnect(f, nsA, "client-a", service, 80)
+			})
```
nitpicky, but the order here is weird: I'd expect nsA-client-a to be created before nsA-client-b. (Likewise with the checks after you create the NetworkPolicy).

Having a consistent order makes it easier to see what the tests are checking and verify that they're correct when skimming the code

tathagatachowdhury replied:
Updated the sequence of the tests.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +640,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			defer cleanupNetworkPolicy(f, policyAllowOnlyFromClientB)
+
+			// client-a should not be able to contact server as the policy is to allow only client-b
+			ginkgo.By("Creating client-a which should not be able to contact the server.", func() {
```
The comment doesn't really add anything that the `By` doesn't already say. (Likewise with the other comments before `By` lines later in this test and other tests.)

tathagatachowdhury replied:
Removed all comments before ginkgo.By() in my code.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +640,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+				testCannotConnect(f, f.Namespace, "client-a", service, 80)
+			})
+			// client-b should be able to contact server as the policy is to allow all
+			ginkgo.By("Creating client-b which should be able to contact the server.", func() {
```
(and this comment is actually incorrect; the "allow all" policy doesn't exist yet)

tathagatachowdhury replied:
Removed the comment.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +640,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+					Name: "allow-all",
+				},
+				Spec: networkingv1.NetworkPolicySpec{
+					// Apply this policy to allow all pods
```
"Apply this policy all pods" (the word "allow" shouldn't be there; this part of the policy isn't allowing or denying anything, it's just specifying what the policy applies to).
(Likewise in the corresponding egress test below)

tathagatachowdhury replied:
Updated the comment.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +640,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			})
+		})
+
+		ginkgo.It("should enforce multiple ingress policies with ingress allow-all policy taking precedence based on PodSelector [Feature:NetworkPolicy]", func() {
```
"based on PodSelector" doesn't make sense as part of the test description here since the second policy uses an empty pod selector. (Likewise for the egress test that follows.)

tathagatachowdhury replied:
Updated the comments.

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +640,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+
+			ginkgo.By("Creating client-a which should not be able to contact the server.", func() {
+				testCannotConnect(f, f.Namespace, "client-a", service, 80)
+			})
```
For completeness, delete `policyDenyAll` now and confirm that client-a *can* connect again.

tathagatachowdhury replied:
Added the test after deleting the `policyDenyAll `

danwinship commented on test/e2e/network/network_policy.go:
```diff
@@ -532,12 +640,601 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+		})
+	})
+
+	ginkgo.Context("NetworkPolicy between pods, each containing a server container and a client container", func() {
```
I'm still dubious about this test, and there is a push on to use fewer images and fewer test programs (eg, to make tests work on Windows pods, which this test case certainly wouldn't). So I think you should just remove this for now. (And remove `createPodWithClientServerAndService` along with it.)

tathagatachowdhury replied:
I have changed the implementation of this test. I have removed the `createPodWithClientServerAndService `.

danwinship reviewed (changes requested):
This is mostly ready (other than needing to be rebased). A bunch of nitpicks, one small additional test suggestion, and one test that I think needs to go.

tathagatachowdhury commented:
/retest

Commit:
Added code for e2e tests for network policy.

tathagatachowdhury commented:
@danwinship: I have rebased the PR.

danwinship commented:
sorry for dropping this again...
/lgtm
/approve
